### PR TITLE
chore: add cargo-deny to codebase

### PR DESCRIPTION
**Pull request by ntn-x2**
Fixes https://github.com/KILTprotocol/ticket/issues/3192 by adding [cargo-deny](https://github.com/EmbarkStudios/cargo-deny) and fixing the two yanked dependencies spotted by the tool.

# TODO

Add the new job to the list of required status checks in the repo settings once the PR is merged.

### CODE REVIEW

File: .cargo-deny.toml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  { id = "RUSTSEC-2022-0080", reason = "`parity-util-mem` is unmaintained but a Substrate dependency. Re-verify upon next polkadot-sdk updates." },
+  { id = "RUSTSEC-2024-0336", reason = "`rustls` vulnerability. Re-verify upon next polkadot-sdk updates." },
+  { id = "RUSTSEC-2024-0344", reason = "`curve25519-dalek` vulnerability. Re-verify upon next polkadot-sdk updates." },
+  { id = "RUSTSEC-2024-0370", reason = "`proc-macro-error` is unmaintained but a Substrate dependency. Re-verify upon next polkadot-sdk updates." },
+]
+yanked = "deny"
+
+[bans]
+allow-wildcard-paths = true
+multiple-versions    = "allow"
```
Comment (Ad96el) [inline .cargo-deny.toml:23]: Oh, that's cool.

File: .github/workflows/check-code.yml [hunk 2]
```diff
@@ -96,6 +96,23 @@ jobs:
...
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+
+      - name: Run `cargo deny`
+        uses: EmbarkStudios/cargo-deny-action@v2
+        with:
+          # All is set in the config file
+          arguments:
+          command-arguments: "--hide-inclusion-graph -c .cargo-deny.toml"
```
[the remark this answers is not in the document]
Comment (ntn-x2) [reply .github/workflows/check-code.yml]: [Now](https://github.com/KILTprotocol/kilt-node/pull/732/commits/b86854cb97a22aaa7e68107b068f2fc08a48a5d5) it does 😁 Each command has possible [diagnostics output](https://embarkstudios.github.io/cargo-deny/checks/advisories/diags.html). We can force-deny those.

[the remark this answers is not in the document]
Comment (ntn-x2) [reply .github/workflows/check-code.yml]: Even better, we can deny all warnings: https://github.com/KILTprotocol/kilt-node/pull/732/commits/15b1d435f99e4ac2ee093480b4b326caf2dc2253.

File: .cargo-deny.toml
Comment (Ad96el) [inline .cargo-deny.toml:1]: Is there a reason, not to use the default name? Would slightly simplify the CLI usage.

File: .github/workflows/check-code.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ad96el) [reply .github/workflows/check-code.yml]: :raised_hands:

File: .cargo-deny.toml
Comment (ntn-x2) [reply .cargo-deny.toml:1]: Yeah I mean `deny.toml` is not really a self-explicative name...

**Review by Ad96el — APPROVED** [review]

**Pull request merged**